### PR TITLE
Add safe cast (not optimized)

### PR DESCRIPTION
**Pull request by MerlinEgalite**
Actually, this PR is a bit disappointing since it does bring any gas improvements when it's not reverting 🥲

Perhaps someone else has good ideas

### CODE REVIEW

**Comment by Rubilmax** [comment]
I think the fact that there's no gas improvement comes down to the fact that Solidity type casting already translates to a sequence of opcodes that will be difficult to simplify further:
- when casting to a smaller type: it's only statically casting the variable (meaning there's no runtime behavior changes to the memory or the stack)
- when casting to a larger type: the bytes which were not used before are zero-ed out

Considering this, I'm against merging this PR

**Comment by MerlinEgalite** [comment]
> I think the fact that there's no gas improvement comes down to the fact that Solidity type casting already translates to a sequence of opcodes that will be difficult to simplify further:
> 
> * when casting to a smaller type: it's only statically casting the variable (meaning there's no runtime behavior changes to the memory or the stack)
> * when casting to a larger type: the bytes which were not used before are zero-ed out
> 
> Considering this & the fact the these libraries are to be used inside Solidity contracts, I'm against merging this PR

It's not to be merged. This is just to showcase that it seems impossible to optimize.

Your answer is a response only for the casting case. The check could have been improved as we're doing the in other libraries. However, this is not the case as the compiler seems to already use the most optimized way to do it.

File: src/math/SafeCast.sol [hunk 1]
```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: GNU AGPLv3
+pragma solidity ^0.8.0;
+
+/// @title WadRayMath.
+/// @author Morpho Labs.
+/// @custom:contact security@morpho.xyz
+/// @notice TODO
+library SafeCast {
+
+	uint256 internal constant MAX_UINT128 = 2**128 - 1; // Not possible to use type(uint128).max in yul
```
Comment (MathisGD) [inline src/math/SafeCast.sol:10]: Why is the tab so l a r g e ?

Comment (MerlinEgalite) [reply src/math/SafeCast.sol:10]: the tab? Same as other repors no?

Comment (MathisGD) [reply src/math/SafeCast.sol:10]: Ok I think that it was a bug on my side

File: src/math/SafeCast.sol [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+				revert(0, 0)
+			}
+
+			casted := value
+		}
+	}
+
+	function toUint8(uint256 value) internal pure returns(uint8 casted) {
+		assembly {
+			if lt(MAX_UINT8, value) {
```
Comment (MathisGD) [inline src/math/SafeCast.sol:69]: Issue here I think

Comment (MathisGD) [reply src/math/SafeCast.sol]: The issue is still there I think

**Comment by MathisGD** [comment]
Should we close this one ?

**Pull request closed without merging**